src/vvwe.rs

```diff
@@ -80,65 +84,69 @@ pub trait CausalOp<A> {
 }
 
 impl<A: Actor, T: CausalOp<A>> CausalityBarrier<A, T> {
-    pub fn new(site_id: A) -> Self {
-        CausalityBarrier { peers: HashMap::new(), buffer: HashMap::new(), local_id: site_id }
+    pub fn new() -> Self {
+        CausalityBarrier {
+            peers: HashMap::new(),
+            buffer: HashMap::new(),
+        }
     }
 
-    pub fn ingest(&mut self, msg: T) -> Option<T> {
-        let v = self.peers.entry(msg.dot().actor).or_insert_with(VectorEntry::new);
-        // Have we already seen this message?
-        if v.is_ready(&msg.dot().counter) {
-            return None
+    pub fn ingest(&mut self, op: T) -> Option<T> {
+        let v = self.peers.entry(op.dot().actor).or_default();
+        // Have we already seen this op?
+        if v.is_ready(op.dot().counter) {
+            return None;
         }
 
-        v.increment(msg.dot().counter);
+        v.increment(op.dot().counter);
 
         // Ok so it's an exception but maybe we can still integrate it if it's not constrained
         // by a happens-before relation.
         // For example: we can always insert into most CRDTs but we can only delete if the
         // corresponding insert happened before!
-        match msg.happens_after() {
-            // Dang! we have a happens before relation!
+        match op.happens_after() {
+            // Dang! we have a happens after relation!
             Some(dot) => {
                 // Let's buffer this operation then.
-                if ! self.saw_site_do(&dot.actor, &dot.counter) {
-                    self.buffer.insert(dot, msg);
-                // and do nothing
+                if !self.saw_site_dot(&dot) {
+                    self.buffer.insert(dot, op);
+                    // and do nothing
                     None
                 } else {
-                    Some(msg)
+                    Some(op)
                 }
             }
             None => {
                 // Ok so we're not causally constrained, but maybe we already saw an associated
                 // causal operation? If so let's just delete the pair
-                match self.buffer.remove(&msg.dot()) {
-                    Some(_) => None,
-                    None => Some(msg),
+                match self.buffer.remove(&op.dot()) {
+                    Some(_) => None, // we are dropping the dependent op! that can't be right
```

Comment:
thats correct but only for a situation where messages are inserts and updates. 

If we had a `Delete(IX)` in the buffer and we see the matching `Insert(IX, C)` then they cancel out.

Reply:
I think we could generalize this by having some sort of multiplication for operations. (Do CRDT operations necessarily form a semigroup?)

Reply:
I don't think cancelling out is a concern of the causality barrier. My intuition is just to keep the barrier as a sort of buffer to re-order and release operations as they are ready to be consumed by the underlying causal algorithm.

Reply:
Consider that the underlying algorithm already  has the logic in place to "cancel" ops if that's how they should behave. 

ie. the LSEQ understands how to "cancel" out an "insert" with a "delete", this logic is in the `apply` method.

To also have this logic in the CausalityBarrier would be duplicating logic and potentially causing bugs if the cancellation logic is not in sync with the underlying algorithm.

Reply:
In the current implementation where causal pairs are exactly insertions and deletions i think it's fine / correct but if we want to add support for batch operations / more complex relations then it'll have to move to vec<op> for sure anyways and cancelling out inside the VVwE is probably not a great idea anymore.